presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -416,12 +447,54 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                     failureDetector,
                     nodeTaskMap,
                     stageSchedulers,
-                    stageLinkages);
+                    stageLinkages,
+                    stagedDependencies,
+                    schedulingQueue);
             stages.addAll(subTree);
 
             SqlStageExecution childStage = subTree.get(0);
             childStagesBuilder.add(childStage);
         }
+
+        for (Reference<StageExecutionPlan> dependency : plan.getDependencies()) {
+            PlanFragmentId fragmentId = dependency.get().getFragment().getId();
+            if (stagedDependencies.contains(fragmentId)) {
```

Comment:
I feel this might relate to avoid executing the same `SuperStage` twice? But I don't quite understand...